controllers/alert_controller.go

```diff
@@ -21,146 +21,180 @@ import (
 
-	var alert v1beta1.Alert
-	if err := r.Get(ctx, req.NamespacedName, &alert); err != nil {
+	alert := &v1beta1.Alert{}
+	if err := r.Get(ctx, req.NamespacedName, alert); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
 	// record suspension metrics
-	r.recordSuspension(ctx, alert)
+	r.RecordSuspend(ctx, alert, alert.Spec.Suspend)
 
-	// record reconciliation duration
-	if r.MetricsRecorder != nil {
-		objRef, err := reference.GetReference(r.Scheme, &alert)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		defer r.MetricsRecorder.RecordDuration(*objRef, reconcileStart)
+	if alert.Spec.Suspend {
+		log.Info("Reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
 	}
 
-	// validate alert spec and provider
-	if err := r.validate(ctx, alert); err != nil {
-		meta.SetResourceCondition(&alert, meta.ReadyCondition, metav1.ConditionFalse, meta.ReconciliationFailedReason, err.Error())
-		if err := r.patchStatus(ctx, req, alert.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
-		}
-		return ctrl.Result{Requeue: true}, err
+	patchHelper, err := patch.NewHelper(alert, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
 	}
 
-	if !apimeta.IsStatusConditionTrue(alert.Status.Conditions, meta.ReadyCondition) || alert.Status.ObservedGeneration != alert.Generation {
-		meta.SetResourceCondition(&alert, meta.ReadyCondition, metav1.ConditionTrue, v1beta1.InitializedReason, v1beta1.InitializedReason)
-		alert.Status.ObservedGeneration = alert.Generation
-		if err := r.patchStatus(ctx, req, alert.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
+	defer func() {
+		conditions.SetSummary(alert, meta.ReadyCondition)
+
+		patchOpts := []patch.Option{
+			patch.WithOwnedConditions{
+				Conditions: []string{
+					meta.ReadyCondition,
+					meta.ReconcilingCondition,
+					meta.StalledCondition,
+				},
+			},
 		}
-		log.Info("Alert initialised")
-	}
 
-	r.recordReadiness(ctx, alert)
+		if retErr == nil && (result.IsZero() || !result.Requeue) {
+			conditions.Delete(alert, meta.ReconcilingCondition)
 
-	return ctrl.Result{}, nil
-}
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
 
-func (r *AlertReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.Alert{}).
-		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicates.ReconcileRequestedPredicate{})).
-		Complete(r)
+			readyCondition := conditions.Get(alert, meta.ReadyCondition)
+			switch readyCondition.Status {
+			case metav1.ConditionFalse:
+				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
+				conditions.MarkStalled(alert, readyCondition.Reason, readyCondition.Message)
+			case metav1.ConditionTrue:
+				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
+				conditions.Delete(alert, meta.StalledCondition)
+			}
+		}
```

Comment:
You may want to have a look and see if the above works in all scenarios, as Michael figured it didn't in some: https://github.com/fluxcd/pkg/pull/144